### PR TITLE
Implement plugin install and uninstall functionality

### LINKED ISSUE #2317: singularity plugin command for administration of plugins
In order to provide an easy way for system administrators to manage the plugins installed on their system, a `singularity plugin` command will be implemented. This command will contain several subcommands related to plugin management:

- `enable`/`disable`
- `install`/`uninistall`
- `configure`
- `list`
- `info`

etc...

### LINKED ISSUE #2321: Plugin install and uninstall commands
Plugin administration commands:

- `singularity plugin install <plugin path>`
- `singularity plugin uninstall <name>`

Still some questions to be answered: 

- Should install take a file and name as arguments, or just a file and have the plugin's name be stored inside of the plugin SIF file?
- Should we reference the SIF file name or plugin name when doing `singularity plugin uninstall`?

### LINKED ISSUE #2324: Plugin list command
`singularity plugin list`

Provides a list of names of installed plugins

### PR DESCRIPTION
**Pull request by mem**
Implement plugin install / uninstall and list subcommands

- Fixes #2321 
- Fixes #2324

### CODE REVIEW

File: cmd/internal/cli/plugin_install_linux.go
Comment (bauerm97) [inline cmd/internal/cli/plugin_install_linux.go]: Similar to what I changed in `PluginCompileCmd`, I think we should just use `Run` here, and do `sylog.Fatalf` instead of printing the error.

File: examples/plugins/test-plugin/main.go
Comment (bauerm97) [inline examples/plugins/test-plugin/main.go:21]: +1

File: internal/app/singularity/plugin_install_linux.go
Comment (bauerm97) [inline internal/app/singularity/plugin_install_linux.go]: Stray debug output left behind here?

File: internal/app/singularity/plugin_install_linux.go
Comment (bauerm97) [inline internal/app/singularity/plugin_install_linux.go:20]: I think in general we should add clarifying information to `return err` statements to make debugging easier

`return fmt.Errorf("while loading sif: %s", err)`

**Review by bauerm97 — CHANGES_REQUESTED** [review]
So for the most part this is really awesome. I think you perfectly captured my intentions when I was writing the first draft of this, hopefully I left detailed enough comments throughout that it wasn't too nasty to take care of ;)

Just a few minor tweaks and tune ups before I think it's ready to merge. We can either wait and add testing to this PR; or (my preference) we can merge the PR as is and create a new issue detailing what testing is required before a `3.1.0` release. I think the second approach is best because there is actually a whole bunch of testing we need to add which spans more than just installation, but I leave it to you to consider

**Comment by bauerm97** [comment]
p.s. I updated the PR message to reference the correct issue. #2317 is an epic which encompasses multiple commands such as `configure`, `compile`, etc... We still will have a few more to go after this PR

File: internal/app/singularity/plugin_install_linux.go
Comment (mem) [reply internal/app/singularity/plugin_install_linux.go:20]: Sure, I still need to figure out how to test some of the error paths.

**Comment by bauerm97** [comment]
Also, I just noticed that we're missing `uninstall` command here

File: cmd/internal/cli/plugin_install_linux.go
Comment (mem) [reply cmd/internal/cli/plugin_install_linux.go]: Yep, I need to go over error reporting for all these commands

File: cmd/internal/cli/plugin_uninstall_linux.go
Comment (bauerm97) [inline cmd/internal/cli/plugin_uninstall_linux.go:20]: Comment incorrectly formed as pointed out by linter

File: internal/pkg/plugin/binary.go
Comment (bauerm97) [inline internal/pkg/plugin/binary.go:117]: Needs a comment as pointed out by linter. Pesky thing

File: internal/pkg/plugin/binary.go
Comment (bauerm97) [inline internal/pkg/plugin/binary.go:388]: I commented this somewhere else, but maybe we can return `map[string]*Meta`, unless we see a reason not to. Key can be the SHA name of the meta file, since it's guaranteed to be unique per name, we shouldn't have any hash map collisions. What're your thoughts here @mem ? I don't see any immediate drawbacks to using a map, can you think of any?

File: internal/pkg/plugin/binary.go
Comment (bauerm97) [inline internal/pkg/plugin/binary.go:120]: Is it possible to change `List()` to return a `map[string]*Meta` to avoid having to iterate over a slice every time? I know it will likely never be an issue because of how quick Go is and the relatively few amount of plugins we expect to be installed

**Review by bauerm97 — CHANGES_REQUESTED** [review]

File: internal/pkg/plugin/binary.go
Comment (mem) [reply internal/pkg/plugin/binary.go:117]: oops, sorry, missed this one, done.

File: cmd/internal/cli/plugin_uninstall_linux.go
Comment (bauerm97) [inline cmd/internal/cli/plugin_uninstall_linux.go]: `PluginUninstall`

File: cmd/internal/cli/plugin_install_linux.go
Comment (bauerm97) [inline cmd/internal/cli/plugin_install_linux.go]: Let's use `sylog.Fatalf()` here instead of `fmt.Printf()`. This way we get automatic exit code handling

Comment (bauerm97) [reply cmd/internal/cli/plugin_install_linux.go]: Same for the rest of the `Run` functions

**Review by bauerm97 — APPROVED** [review]
Assuming CI passes, this looks good to me.

**Pull request merged**
